### ISSUE TITLE
Refactor to Have a Decoupled "Generator" from Decoder

### ISSUE THREAD

**Issue by shuoyangd** [pytorch/fairseq#319]
Hi All,

I just started using fairseq recently. Thanks for the work! It's been pretty pleasant to use so far. 

I'd like propose a refactor idea that introduces a separated "generator" class (following OpenNMT naming convention) to represent the final linear transformation and softmax other than integrating them to the decoder. There are several reasons for this:

1. They are doing roughly the same thing and can potentially be unified into one class. Especially, as you are in the middle of adding adaptive softmax, having a unified generator allows you to only change the code once instead of three times.
2. Many research ideas only involves changing "generator", e.g. [pointer generator](https://arxiv.org/abs/1704.04368), [target-side syntax](https://arxiv.org/abs/1702.03525). A stand-alone "generator" class avoids the need to mess with the decoder (especially the beam search part) altogether when implementing these ideas.
3. Right now LSTM, FConv and Transformer each has a slightly different naming convention and implementation for this part, which is a bit messy if you look at it. Having a separate class will improve readability of the code.

I have a working branch for this (https://github.com/shuoyangd/fairseq/tree/dev). Right now I can somewhat reproduce the behavior of LSTM, FConv and Transformer translation model with this modification (exactly the same training/dev loss after 1st epoch on IWSLT de-en). I'm wondering if this is something that you may potentially want to merge. If yes, I can do some more work to ensure it lives up to the standard of a pull request.

I also understand that the OpenNMT naming convention might not be a good fit for this repo as it is used for something else (`SequenceGenerator`) and I'm open to naming suggestions for this class.

Thanks!

Shuoyang
